File: backend/schemas/offboarding.py
"""
Schemas Pydantic para offboarding — instancias y activos.
"""
from datetime import date
from typing import List, Literal, Optional
from uuid import UUID

from pydantic import BaseModel

MotivoEgreso = Literal[
    "renuncia", "despido", "acuerdo_mutuo",
    "fin_contrato", "jubilacion", "fallecimiento", "otro",
]


class OffboardingCreate(BaseModel):
    empleado_id: UUID
    motivo: MotivoEgreso
    fecha_ultimo_dia: Optional[date] = None
    descripcion_motivo: Optional[str] = None


class EfectivizarBaja(BaseModel):
    """Body de `POST /api/offboarding/{instancia_id}/efectivizar`: la baja EFECTIVA del empleado.

    Un solo campo, y es a propósito. `fecha_egreso` es el HECHO —el día que la persona
    efectivamente dejó de trabajar— y no tiene por qué coincidir con la previsión que se cargó al
    abrir el trámite (`offboarding_instancias.fecha_ultimo_dia`). **Que difieran NO es un error y
    el service no las sincroniza**: la previsión queda como quedó, que es lo que permite después
    comparar lo previsto con lo ocurrido.

    El motivo NO va acá: ya se eligió al abrir el proceso y vive en `motivo_egreso`. Pedirlo de
    nuevo abriría la puerta a dos motivos distintos para la misma salida.
    """
    fecha_egreso: date


class ActivoUpdate(BaseModel):
    devuelto: bool


class EntrevistaUpdate(BaseModel):
    """Entrevista de salida. Las notas son opcionales: se puede marcar que la entrevista se
    hizo sin dejar constancia escrita, y también guardar notas de una entrevista en curso."""
    entrevista_salida: bool
    notas_entrevista: Optional[str] = None


class ActivoResponse(BaseModel):
    id: UUID
    tipo_activo: str
    descripcion: Optional[str] = None
    estado: str
    devuelto: bool


class AccesoResponse(BaseModel):
    id: UUID
    tipo: str
    descripcion: Optional[str] = None
    revocado: bool


class OffboardingResponse(BaseModel):
    id: UUID
    empleado_id: UUID
    empresa_id: Optional[UUID] = None
    empresa_nombre: Optional[str] = None
    empleado_nombre: str
    motivo: str
    estado: str
    fecha_inicio: str
    progreso: int
    entrevista_salida: bool = False
    notas_entrevista: Optional[str] = None
    activos: List[ActivoResponse] = []
    accesos: List[AccesoResponse] = []

File: frontend/services/offboarding.ts
import { apiFetch, descargarArchivo, type FormatoExport } from "@/services/api"
import type { OffboardingCreate, OffboardingInstancia } from "@/types/offboarding"

export async function fetchOffboardings(): Promise<OffboardingInstancia[]> {
  return apiFetch<OffboardingInstancia[]>("/api/offboarding")
}

/**
 * Exporta los offboardings ACTIVOS — los mismos que muestra la pantalla.
 *
 * Ninguna de las dos puntas manda query params y las dos mandan `X-Empresa-Id` por
 * `authHeaders()`, así que traen el mismo conjunto. 🔴 El día que el listado gane un filtro
 * (motivo, estado), los dos tienen que armar sus params con UNA función de traducción
 * compartida — molde: `queryVacantes` en services/vacantes.ts.
 */
export function exportarOffboardings(formato: FormatoExport): Promise<void> {
  return descargarArchivo("/api/offboarding/exportar", formato, "offboardings")
}

export async function iniciarOffboarding(data: OffboardingCreate): Promise<OffboardingInstancia> {
  return apiFetch<OffboardingInstancia>("/api/offboarding", {
    method: "POST",
    body: JSON.stringify(data),
  })
}

export async function marcarActivoDevuelto(
  instanciaId: string,
  activoId: string,
  devuelto: boolean,
): Promise<void> {
  await apiFetch<{ ok: boolean }>(
    `/api/offboarding/${instanciaId}/activos/${activoId}`,
    { method: "PUT", body: JSON.stringify({ devuelto }) },
  )
}

/**
 * Registra la entrevista de salida de un offboarding.
 * empresa_id NO va como parámetro: apiFetch inyecta X-Empresa-Id y el backend acota la
 * instancia a esa empresa, igual que el resto de las escrituras del módulo.
 */
export async function registrarEntrevista(
  instanciaId: string,
  entrevistaSalida: boolean,
  notas: string | null,
): Promise<void> {
  await apiFetch<{ ok: boolean }>(`/api/offboarding/${instanciaId}/entrevista`, {
    method: "PUT",
    body: JSON.stringify({ entrevista_salida: entrevistaSalida, notas_entrevista: notas }),
  })
}

/**
 * Efectiviza la baja: escribe `fecha_egreso` en el empleado y cierra la instancia.
 *
 * 🔴 ES EL ÚNICO CAMINO QUE DA DE BAJA A ALGUIEN, y es irreversible desde la UI. Abrir el
 * trámite (`iniciarOffboarding`) ya NO cambia el estado del empleado: hasta que esto corre, la
 * persona sigue contando en headcount, organigrama, ausentismo y saldo de vacaciones.
 *
 * `fecha_egreso` es el HECHO —el día que la persona dejó de trabajar— y no se sincroniza con la
 * previsión que se cargó al abrir el trámite (`fecha_ultimo_dia`). Que difieran no es un error:
 * es lo que después permite comparar lo previsto con lo ocurrido.
 *
 * 🔴 LOS CUATRO ERRORES SE MUESTRAN CON SU MENSAJE, sin traducir. Los dos de fecha
 * —FECHA_EGRESO_FUTURA y FECHA_EGRESO_INVALIDA (anterior al ingreso)— son los que alguien
 * encuentra operando normal, y cada uno dice qué corregir. Los otros dos son
 * OFFBOARDING_YA_CERRADO y EMPLEADO_YA_DE_BAJA (409), que aparecen con dos pestañas abiertas.
 */
export async function efectivizarBaja(instanciaId: string, fechaEgreso: string): Promise<void> {
  await apiFetch<{ ok: boolean }>(`/api/offboarding/${instanciaId}/efectivizar`, {
    method: "POST",
    body: JSON.stringify({ fecha_egreso: fechaEgreso }),
  })
}
